### DISCUSSION TITLE
Issue with "Method Not Allowed" error when creating bucket using Ozone S3 Gateway

### DISCUSSION THREAD

**Discussion by FaheemKhan0817** [apache/ozone#6662] [FAQ] [answered]
#--->> replace with actual credentials 

I hope this message finds you well. I'm currently facing an issue while working with the Ozone S3 Gateway. When attempting to create a bucket using the create_bucket operation, I encounter a "Method Not Allowed" error.

Here's a snippet of the code I'm using:

import boto3

# Create an S3 client with the Ozone S3 gateway endpoint
s3 = boto3.client('s3', endpoint_url='http://52.4.108.136:9888',
                  aws_access_key_id='####################',
                  aws_secret_access_key='#############################')

# Create a bucket within the existing volume
volume_name = 'vol1'
bucket_name = 'my-new-bucket'
response = s3.create_bucket(Bucket=f's3v/{volume_name}/{bucket_name}')
print(f"Bucket {bucket_name} created: {response['ResponseMetadata']['HTTPStatusCode']}")

ozone is running on AWS instance  and i used Hadoop vms of my company

Could you please assist me in troubleshooting this issue? Any insights or guidance would be greatly appreciated.

Thank you for your time and assistance.

**Comment by adoroszlai** [comment, marked as answer]
Hi @FaheemKhan0817,

Do not include volume name and/or `s3v` in bucket to be created.  Volume name is internal to Ozone, S3 clients are not aware of it.

> `response = s3.create_bucket(Bucket=f's3v/{volume_name}/{bucket_name}')`

It should be something like:

`response = s3.create_bucket(Bucket=bucket_name)`

**Reply by FaheemKhan0817** [reply]
@adoroszlai    still same issue 
PS C:\Practice> & C:/Users/hp/AppData/Local/Programs/Python/Python311/python.exe c:/Practice/ozone_test.py
Traceback (most recent call last):
  File "c:\Practice\ozone_test.py", line 32, in <module>
    response = s3.create_bucket(Bucket='s3v')
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\site-packages\botocore\client.py", line 565, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\site-packages\botocore\client.py", line 1021, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (405) when calling the CreateBucket operation: Method Not Allowed
PS C:\Practice>

**Reply by adoroszlai** [reply]
> `s3 = boto3.client('s3', endpoint_url='http://52.4.108.136:9888/',`

Port 9888 is most likely also wrong, it's Recon's default port.  Try 9878, which is S3 Gateway's default port.

**Reply by FaheemKhan0817** [reply]
http://52.4.103.135:9888/#/Buckets    When I click on this linke I cleary see three already created bucket  on Apache Osone Recon UI

**Reply by adoroszlai** [reply]
> When I click on this linke I cleary see three already created bucket on Apache Osone Recon UI

But S3 client should not connect to Recon, rather S3 Gateway.  So please use the latter's address.

**Reply by FaheemKhan0817** [reply]
Can i used stand-alone Ozone for my work without using s3 and Aws through Python3?

**Reply by spacemonkd** [reply]
For Python3 Ozone doesn't have a native client. Only Java client is present.
Refer to: https://github.com/apache/ozone/discussions/6665#discussioncomment-9387366
